src/appleseed/renderer/modeling/volume/genericvolume.cpp

```diff
@@ -141,32 +152,89 @@ class GenericVolume
+            values->m_precomputed.m_albedo[i] = values->m_scattering[i] == 0.0f ? 0.0f :
+                values->m_scattering[i] / values->m_precomputed.m_extinction[i];
+        }
     }
 
-    float sample(
-        SamplingContext&    sampling_context,
-        const void*         data,
-        const ShadingRay&   volume_ray,
-        const float         distance,
-        Vector3f&           incoming) const override
+    void sample(
+        foundation::Arena&      arena,
+        SamplingContext&        sampling_context,
+        const void*             data,
+        DistanceSample&         sample) const override
     {
-        sampling_context.split_in_place(2, 1);
-        const Vector2f s = sampling_context.next2<Vector2f>();
+        const InputValues* values = static_cast<const InputValues*>(data);
+        const Spectrum& extinction_coef = values->m_precomputed.m_extinction;
+
+        // Sample channel.
+        sampling_context.split_in_place(1, 1);
+        const float s0 = sampling_context.next2<float>();
+        size_t channel = 0;
+        float cdf = sample.m_channel_sampling_weights[0];
+        while (s0 > cdf && channel + 1 < Spectrum::size())
+        {
+            cdf += sample.m_channel_sampling_weights[++channel];
+        }
+
+        // Sample distance exponentially, assuming that the ray is infinite.
+        sampling_context.split_in_place(1, 1);
+        const float s1 = sampling_context.next2<float>();
+        const double distance = sample_exponential_distribution(s1, extinction_coef[channel]);
 
-        const Vector3f outgoing(normalize(volume_ray.m_dir));
-        return m_phase_function->sample(outgoing, s, incoming);
+        // Evaluate this sample.
+        evaluate(arena, data, distance, sample);
     }
 
-    float evaluate(
-        const void*         data,
-        const ShadingRay&   volume_ray,
-        const float         distance,
-        const Vector3f&     incoming) const override
+    void evaluate(
+        foundation::Arena&      arena,
+        const void*             data,
+        const double            distance,
+        DistanceSample&         sample) const override
     {
-        const Vector3f outgoing = Vector3f(normalize(volume_ray.m_dir));
-        return m_phase_function->evaluate(outgoing, incoming);
+        const InputValues* values = static_cast<const InputValues*>(data);
+
+        const Spectrum& extinction_coef = values->m_precomputed.m_extinction;
+        const ShadingRay& volume_ray = sample.m_incoming_point->get_ray();
+
+        sample.m_distance = distance;
+        sample.m_probability = 0.0f;
+
+        if (volume_ray.is_finite() && sample.m_distance > volume_ray.get_length())
+        {
+            // The ray is transmitted.
+            sample.m_distance = volume_ray.get_length();
+            evaluate_transmission(data, volume_ray, sample.m_value);
+            sample.m_transmitted = true;
+            for (int i = 0; i < Spectrum::size(); ++i)
```

Comment:
Use `size_t`, this won't compile on Windows.

Reply:
In fact, you should do as you did in other similar loops:

    for (size_t i = 0, e = Spectrum::size(); i < e; ++i)